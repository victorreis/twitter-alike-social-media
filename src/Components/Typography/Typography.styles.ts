import styled from 'styled-components';

import { TypographyStyleProps } from './Typography.types';

export const TypographyContainer = styled.span<TypographyStyleProps>`
  color: ${({ theme }) => theme.colors.font.default};
  ${({ variant, theme }) => theme.typographies[variant]};
`;
