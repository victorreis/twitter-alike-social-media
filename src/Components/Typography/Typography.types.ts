import { TestProps } from '../../Config/Tests/Test.types';
import { TypographyVariant } from '../../Theme/Types/Typographies.types';

export interface RequiredTypographyProps {
  /**
   * Component's children.
   */
  children: React.ReactNode;
}

export interface DefaultTypographyProps {
  /**
   * Sets the component variant. It changes the HTML tag and the styles.
   * @default 'body1'.
   */
  variant?: TypographyVariant;
}

export interface OptionalTypographyProps {}

export type TypographyProps = RequiredTypographyProps &
  DefaultTypographyProps &
  OptionalTypographyProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLSpanElement>, 'children'>;

export type TypographyStyleProps = Required<DefaultTypographyProps>;
