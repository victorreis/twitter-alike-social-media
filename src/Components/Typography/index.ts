export * from './Typography.types';
export * from './Typography';
