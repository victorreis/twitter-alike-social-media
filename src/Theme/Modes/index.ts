export { darkTheme } from './Dark.theme';
