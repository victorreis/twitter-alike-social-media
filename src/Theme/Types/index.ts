export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
