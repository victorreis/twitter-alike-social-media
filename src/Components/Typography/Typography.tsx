import { TestProps } from '../../Config/Tests/Test.types';
import { typographyVariantToTag } from '../../Theme/Types/Typographies.types';
import { TypographyContainer } from './Typography.styles';
import { TypographyProps, DefaultTypographyProps } from './Typography.types';

export const typographyDefaults: Required<DefaultTypographyProps> &
  Required<TestProps> = {
  testID: 'Typography',
  variant: 'body1',
};

export const Typography: React.FC<TypographyProps> = (props): JSX.Element => {
  const {
    testID = typographyDefaults.testID,
    children,
    variant = typographyDefaults.variant,
    style,
    ...others
  } = props;

  return (
    <TypographyContainer
      as={typographyVariantToTag[variant]}
      data-testid={testID}
      style={style}
      variant={variant}
      {...others}
    >
      {children}
    </TypographyContainer>
  );
};
