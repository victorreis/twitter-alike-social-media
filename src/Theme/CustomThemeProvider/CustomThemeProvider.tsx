/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
