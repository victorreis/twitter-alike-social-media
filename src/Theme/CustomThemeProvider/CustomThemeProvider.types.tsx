import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
