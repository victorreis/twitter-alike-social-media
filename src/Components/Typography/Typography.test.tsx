import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { Typography, typographyDefaults } from './Typography';
import { RequiredTypographyProps, TypographyProps } from './Typography.types';

describe('typography component tests', () => {
  const text = 'text';
  const newVariant = 'h1';

  const requiredProps: RequiredTypographyProps = {
    children: text,
  };

  const setup = (props?: TypographyProps) => {
    const renderRTR = () =>
      renderRTRCreator<TypographyProps>(Typography, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<TypographyProps>(Typography, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it('should render the component', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(typographyDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it('should render the text', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(text);

      expect(element).toBeInTheDocument();
    });

    it(`should render '${typographyDefaults.variant}' as the default 'variant'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        variant: typographyDefaults.variant,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'variant' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        variant: newVariant,
      }).renderRTR().root;
      const element = instance.findByProps({ variant: newVariant });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it('should have styled the Container component', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(typographyDefaults.testID);

      expect(container).toHaveStyle({
        color: hexToRgb(themes.default.colors.font.default),
        ...themes.default.typographies[typographyDefaults.variant],
      });
    });
  });

  describe('snapshot tests', () => {
    it('should render correctly', () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
