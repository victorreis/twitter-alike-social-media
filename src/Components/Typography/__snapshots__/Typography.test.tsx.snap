// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typography component tests snapshot tests should render correctly 1`] = `
<span
  className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 llowaZ"
  data-testid="Typography"
>
  text
</span>
`;
